shell/src/main/java/alluxio/cli/fs/command/DistributedCpCommand.java

```diff
@@ -59,23 +71,85 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
     String[] args = cl.getArgs();
     AlluxioURI srcPath = new AlluxioURI(args[0]);
     AlluxioURI dstPath = new AlluxioURI(args[1]);
-    Thread thread = CommonUtils.createProgressThread(2L * Constants.SECOND_MS, System.out);
-    thread.start();
-    try {
-      AlluxioConfiguration conf = mFsContext.getPathConf(dstPath);
-      JobGrpcClientUtils.run(new MigrateConfig(srcPath.getPath(), dstPath.getPath(),
-          conf.get(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT), true,
-          false), 1, mFsContext.getPathConf(dstPath));
-    } catch (InterruptedException e) {
-      Thread.currentThread().interrupt();
-      return -1;
-    } finally {
-      thread.interrupt();
+
+    if (PathUtils.hasPrefix(dstPath.toString(), srcPath.toString())) {
+      throw new RuntimeException(ExceptionMessage.MIGRATE_CANNOT_BE_TO_SUBDIRECTORY.getMessage(
+          srcPath, dstPath));
     }
-    System.out.println("Copied " + srcPath + " to " + dstPath);
+
+    mActiveJobs = DEFAULT_ACTIVE_JOBS;
+
+    AlluxioConfiguration conf = mFsContext.getPathConf(dstPath);
+    mWriteType = conf.get(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT);
+
+    distributedCp(srcPath, dstPath);
     return 0;
   }
 
+  private CopyJobAttempt newJob(String srcPath, String dstPath) {
+    CopyJobAttempt jobAttempt = new CopyJobAttempt(mClient,
+        new MigrateConfig(srcPath, dstPath, mWriteType, true, false),
+        new CountingRetry(3));
+
+    jobAttempt.run();
+
+    return jobAttempt;
+  }
+
+  private void distributedCp(AlluxioURI srcPath, AlluxioURI dstPath)
+      throws IOException, AlluxioException {
+    if (mFileSystem.getStatus(srcPath).isFolder()) {
+      createFolders(srcPath, dstPath);
+    }
+    copy(srcPath, dstPath);
```

Comment:
`createFolders` does a listing, and `copy` also does a listing. Can we avoid doing a double recursive listing of the source?

Reply:
I would prefer not because:

1. copy is expensive anyways so listStatus is a small cost here.
2. More intuitive log messages.

Reply:
Could you explain why it is more intuitive with 2 listings, and why 1 listing cannot have good messages?

One concern is that if a listing is a large tree, it would have to lock that large tree twice. Ideally, we would only have to lock that subtree once.